### PR TITLE
feat(store): compute FieldLayout at compile time

### LINKED ISSUE #1327: Encode table schemas and fieldLayout at compile time
Depends on https://github.com/latticexyz/mud/issues/1326
If Schemas are just lengths, it's pretty trivial to make a typescript Schema builder and pre-render a hex string, since table schemas are immutable.
`getSchema` consumes quite a lot of gas (e.g. `2814` for 6 entries) and this would remove all of it

### PR DESCRIPTION
**Pull request by Boffee**
Followup to #1336
Fixes #1327 

Computing FieldLayout bytes from schema field type array is expensive (~500 + 500 * # of schema fields). This change precomputes the FieldLayout offchain inside codegen and hardcodes the byte into the auto-generated table libraries.

### CODE REVIEW

File: packages/cli/contracts/src/codegen/tables/Dynamics1.sol [hunk 1]
```diff
@@ -32,9 +36,7 @@ struct Dynamics1Data {
 library Dynamics1 {
   /** Get the table values' field layout */
   function getFieldLayout() internal pure returns (FieldLayout) {
```
Comment (holic) [inline packages/cli/contracts/src/codegen/tables/Dynamics1.sol:38]: Do we need this method anymore?

File: packages/store/ts/codegen/renderFieldLayout.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  if (dynamicFields > MAX_DYNAMIC_FIELDS) throw new Error(`FieldLayout: invalid length ${dynamicFields}`);
+
+  fieldLayout |= BigInt(totalLength) << BigInt(LayoutOffsets.TOTAL_LENGTH);
+  fieldLayout |= BigInt(staticFields) << BigInt(LayoutOffsets.NUM_STATIC_FIELDS);
+  fieldLayout |= BigInt(dynamicFields) << BigInt(LayoutOffsets.NUM_DYNAMIC_FIELDS);
+
+  return `0x${fieldLayout.toString(16).padStart(64, "0")}`;
+}
```
Comment (alvrs) [inline packages/store/ts/codegen/renderFieldLayout.ts:30]: slightly rearanged and renamed some variables to make it as close as possible to `FieldLayout.sol` (which makes auditing easier)

```suggestion
export function encodeFieldLayout(fields: RenderType[]) {
  const staticFields = fields.filter(({ isDynamic }) => !isDynamic);
  const numDynamicFields = fields.length - staticFields.length;
  
  let fieldLayout = 0n;
  let totalLength = 0;
  const totalFields = fields.length;
  
  if (totalFields > MAX_TOTAL_FIELDS) throw new Error(`FieldLayout: invalid length ${totalFields}`);
  if (numDynamicFields > MAX_DYNAMIC_FIELDS) throw new Error(`FieldLayout: invalid length ${numDynamicFields}`);

  for (let i = 0; i < staticFields.length; i++) {
    const { isDynamic, staticByteLength } = fields[i];
    if (isDynamic) throw new Error(`FieldLayout: static type after dynamic type`);

    totalLength += staticByteLength;
    fieldLayout |= BigInt(staticByteLength) << BigInt((WORD_LAST_INDEX - 4 - i) * BYTE_TO_BITS);
  }

  fieldLayout |= BigInt(totalLength) << BigInt(LayoutOffsets.TOTAL_LENGTH);
  fieldLayout |= BigInt(staticFields.length) << BigInt(LayoutOffsets.NUM_STATIC_FIELDS);
  fieldLayout |= BigInt(numDynamicFields) << BigInt(LayoutOffsets.NUM_DYNAMIC_FIELDS);

  return `0x${fieldLayout.toString(16).padStart(64, "0")}`;
}
```

File: packages/store/ts/codegen/renderFieldLayout.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { RenderType } from "@latticexyz/common/codegen";
+import { BYTE_TO_BITS, LayoutOffsets, MAX_DYNAMIC_FIELDS, MAX_TOTAL_FIELDS, WORD_LAST_INDEX } from "../constants";
+
+export function renderFieldLayout(fields: RenderType[]) {
+  return `FieldLayout constant _fieldLayout = FieldLayout.wrap(${encodeFieldLayout(fields)});`;
+}
+
```
Comment (alvrs) [inline packages/store/ts/codegen/renderFieldLayout.ts:7]:
```suggestion
// Make sure this logic stays aligned with @latticexyz/store/src/FieldLayout.sol
```

File: packages/store/ts/constants.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/* Shared constants */
+
```
Comment (alvrs) [inline packages/store/ts/constants.ts:2]: // Make sure these stay aligned with @latticexyz/store/src/constants.sol

Comment (alvrs) [inline packages/store/ts/constants.ts:2]:
```suggestion
// Make sure these stay aligned with @latticexyz/store/src/constants.sol

```

**Review by alvrs — COMMENTED** [review]
amazing, thank you!!

File: packages/store/ts/codegen/renderFieldLayout.ts (hunk 2 above)
Comment (holic) [reply packages/store/ts/codegen/renderFieldLayout.ts:30]: can we add tests for this too? seems important to get this right (and would be nice to have the same tests both here and in solidity)

we have a version of this in protocol-parser but it's also untested and doesn't match the solidity code as well

File: packages/cli/contracts/src/codegen/tables/Dynamics1.sol (hunk 1 above)
Comment (alvrs) [reply packages/cli/contracts/src/codegen/tables/Dynamics1.sol:38]: seems fine for completeness (so the methods match what you can get from the `Tables` table), if it's not used anywhere it's also not included in the bytecode

**Review by alvrs — APPROVED** [review]

**Review by alvrs — APPROVED** [review]

File: packages/cli/contracts/src/codegen/tables/Dynamics1.sol (hunk 1 above)
Comment (holic) [reply packages/cli/contracts/src/codegen/tables/Dynamics1.sol:38]: I don't know why anyone would choose the less efficient one (function call) over the more efficient one though (importing the `_fieldLayout` directly)

**Pull request merged**